### ISSUE TITLE
Add an option to influence sampler decision based span's StartOption

### ISSUE THREAD

**Issue by Lameorc** [open-telemetry/opentelemetry-go#3636]
### Problem Statement

I'm currently migrating an application from Opencensus where we're using [WithSampler](https://github.com/census-instrumentation/opencensus-go/blob/49838f207d61097fc0ebb8aeef306913388376ca/trace/trace.go#L162) to ignore certain spans dynamically.

Specifically, we create an opencensus StartOption using
```
trace.WithSampler(trace.NeverSample())
```
where `trace` is from `"go.opencensus.io/trace"`

I'm looking for a way to migrate this into the opentelemetry ecosystem and have not managed to find one. I think I can write my custom Option? However this is something I'd expect to be provided by the core library, as it doesn't seem that far fetched.

Furthermore, I've not found this mentioned anywhere in the docs (which are in themselves very barebones, but that's out of scope of this issue).

### Proposed Solution

Implement a new function akin to `WithSpanKind(kind SpanKind) SpanStartOption` (and similiar) that produces a `SpanStartOption` which - when applied - ensures that the Span's sampled flag is always false.

In case you consider this to be too specific, at least consider this for the `contrib` repository/extensions.

#### Alternatives

I've considered two alternatives:
1. Write my own implementation. This is likely what I'll go with to "unblock" the migration. However, this doesn't seem ideal for (what I consider) to be a pretty basic start option.
2. Set a specific attribute on the span, which is then checked by a custom sampler. The sampler decides, based on the attribute whether to sample or not. I've discarded this option fairly quickly as I don't like the coupling and I can't set the attributes on a non-existent span AFAIK. So I would still need to create a (potentially) sampled span and then create a new one just for this. I'm not even sure this whole approach is viable, but it would be my last resort I guess.

#### Prior Art

- [Function producing start option from sampler](https://github.com/census-instrumentation/opencensus-go/blob/49838f207d61097fc0ebb8aeef306913388376ca/trace/trace.go#L162)
- [Function producing Sampler provided to the above function](https://github.com/census-instrumentation/opencensus-go/blob/master/trace/sampling.go#L71)

**Comment by Lameorc** [comment]
Just for posterity. I've solved it using the following snippet in my middleware for now:
```
if t.isPathExcluded(req.URL.Path) { // excluded paths are not traced
	otelCtx := trace.SpanContextFromContext(ctx)
	unsampledCtx := otelCtx.WithTraceFlags(
		otelCtx.TraceFlags().WithSampled(false),
	)
	ctx = trace.ContextWithSpanContext(ctx, unsampledCtx)
}
```

I'm not sure this is the correct approach or if a better one exists. But if this is the proper way to do it, It would stand a mention in docs.

**Comment by MrAlias** [comment]
Similar to https://github.com/open-telemetry/opentelemetry-go/issues/401. As mentioned there, this is an issue that can be solved with a custom sampler as well.

**Comment by MrAlias** [comment]
> Just for posterity. I've solved it using the following snippet in my middleware for now:
> 
> ```go
> if t.isPathExcluded(req.URL.Path) { // excluded paths are not traced
> 	otelCtx := trace.SpanContextFromContext(ctx)
> 	unsampledCtx := otelCtx.WithTraceFlags(
> 		otelCtx.TraceFlags().WithSampled(false),
> 	)
> 	ctx = trace.ContextWithSpanContext(ctx, unsampledCtx)
> }
> ```
> 
> I'm not sure this is the correct approach or if a better one exists. But if this is the proper way to do it, It would stand a mention in docs.

For what it is worth, this will not work if the original `ctx` does not contain a valid span (the trace-id/span-id being zero will mean children do not consider this span context). If it does contain a valid span, that parent sampling state will not be altered based on this operation.

I imagine a custom sampler would better address this situation.
